File: suc/swift/026.swift
import SwiftUI

// MARK: - Color Extension for Hex Initialization
extension Color {
    /// Initializes a Color from a 32-bit hexadecimal integer (0xAARRGGBB or 0xRRGGBB).
    /// If the alpha component is not explicitly provided (i.e., hex value is < 0x1000000),
    /// it defaults to a fully opaque color (alpha = 1.0).
    init(hex: UInt) {
        let red = Double((hex & 0xFF0000) >> 16) / 255.0
        let green = Double((hex & 0x00FF00) >> 8) / 255.0
        let blue = Double(hex & 0x0000FF) / 255.0
        let alpha = Double((hex & 0xFF000000) >> 24) / 255.0

        // If the hex value is in 0xRRGGBB format (no explicit alpha), set alpha to 1.0
        // Otherwise, use the extracted alpha.
        self.init(red: red, green: green, blue: blue, opacity: (hex & 0xFF000000) == 0 ? 1.0 : alpha)
    }
}

// MARK: - AppTokens
struct AppTokens {
    struct Colors {
        static let primary = Color(hex: 0xFF38BDF8)
        static let secondary = Color(hex: 0xFFA78BFA)
        static let tertiary = Color(hex: 0xFF7DD3FC)
        static let background = Color(hex: 0xFFF1F5F9)
        static let surface = Color(hex: 0xFFFFFFFF)
        static let surfaceVariant = Color(hex: 0xFFE2E8F0)
        static let outline = Color(hex: 0xFFCBD5E1)
        static let success = Color(hex: 0xFF22C55E)
        static let warning = Color(hex: 0xFFF59E0B)
        static let error = Color(hex: 0xFFEF4444)
        static let onPrimary = Color(hex: 0xFFFFFFFF)
        static let onSecondary = Color(hex: 0xFF1E1E1E)
        static let onTertiary = Color(hex: 0xFF1E1E1E)
        static let onBackground = Color(hex: 0xFF1E1E1E)
        static let onSurface = Color(hex: 0xFF1E1E1E)
    }

    struct TypographyTokens {
        static let display = Font.system(size: 28, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular) // FontWeight.Normal maps to .regular
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Shapes {
        static let small: CGFloat = 6
        static let medium: CGFloat = 12
        static let large: CGFloat = 20
    }

    struct Spacing {
        static let sm: CGFloat = 8
        static let md: CGFloat = 12
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
        static let xxl: CGFloat = 32
    }

    struct ShadowSpec {
        let elevation: CGFloat // Conceptual elevation, often influences radius and dy
        let radius: CGFloat    // Blur radius of the shadow
        let dy: CGFloat        // Vertical offset of the shadow
        let opacity: Double    // Opacity of the shadow color
    }

    struct ElevationMapping {
        static let level1 = ShadowSpec(elevation: 2, radius: 4, dy: 2, opacity: 0.12)
        static let level2 = ShadowSpec(elevation: 6, radius: 8, dy: 4, opacity: 0.18)
    }
}

// MARK: - RootScreen
struct RootScreen: View {
    // Equivalent to `var scrollOffset by remember { mutableStateOf(Offset(0f, 0f)) }`
    @State private var scrollOffset: CGSize = .zero

    var body: some View {
        // Equivalent to Scaffold's main content layout, with containerColor as background
        VStack(spacing: 0) {
            // Equivalent to CenterAlignedTopAppBar
            Text("Glass Doc Reader")
                .font(AppTokens.TypographyTokens.display)
                .foregroundColor(AppTokens.Colors.onSurface)
                .padding(.top, 20)
                .padding(.vertical, AppTokens.Spacing.lg) // Provides vertical spacing for the title
                .frame(maxWidth: .infinity) // Centers the text horizontally
                .background(AppTokens.Colors.background) // Top bar background color

            // Main content area, equivalent to Compose's Box
            ZStack(alignment: .center) { // contentAlignment = Alignment.Center
                // Background gradient for the Box
                LinearGradient(
                    gradient: Gradient(colors: [
                        AppTokens.Colors.surfaceVariant.opacity(0.8),
                        AppTokens.Colors.background,
                        AppTokens.Colors.surfaceVariant.opacity(0.8)
                    ]),
                    startPoint: .top,
                    endPoint: .bottom
                )
                .frame(maxWidth: .infinity, maxHeight: .infinity) // Fills the available space below the top bar
                
                // Equivalent to Compose's Surface
                VStack(alignment: .leading, spacing: AppTokens.Spacing.md) { // Equivalent to Compose's Column
                    Text("Document Reader")
                        .font(AppTokens.TypographyTokens.title)
                        .foregroundColor(AppTokens.Colors.primary)

                    ForEach(0..<8) { index in
                        Text("This is a sample line of text content number \(index + 1).")
                            .font(AppTokens.TypographyTokens.body)
                            .foregroundColor(AppTokens.Colors.onSurface)
                    }
                    Spacer()
                }
                .padding(AppTokens.Spacing.lg) // Padding inside the Surface content
                .frame(width: UIScreen.main.bounds.width * 0.9, height: 400) // fillMaxWidth(0.9f) and height(400.dp)
                .background(AppTokens.Colors.surface.opacity(0.7)) // color = AppTokens.Colors.surface.copy(alpha = 0.7f)
                .cornerRadius(AppTokens.Shapes.large) // shape = AppTokens.Shapes.large
                .shadow(
                    color: Color.black.opacity(AppTokens.ElevationMapping.level2.opacity),
                    radius: AppTokens.ElevationMapping.level2.radius,
                    x: 0, // Assuming no horizontal offset for shadow as not explicitly defined in Compose's ShadowSpec
                    y: AppTokens.ElevationMapping.level2.dy
                )
                .offset(y: scrollOffset.height / 5) // offset(y = scrollOffset.y.dp / 5)
                .gesture(
                    DragGesture()
                        .onChanged { value in
                            scrollOffset.height += value.translation.height
                        }
                )
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity) // Box fillMaxSize()
        }
        .background(AppTokens.Colors.background.ignoresSafeArea()) // Scaffold containerColor applied to the VStack
    }
}

// MARK: - App Entry Point
@main
struct SingleFileUIApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
                // Ensures the app content extends to all edges of the screen
                .ignoresSafeArea(.all)
                // Hides the top status bar, achieving full-screen display as per requirement
                // Compatible with iOS 13.0 and later.
                .statusBarHidden(true)
        }
    }
}

// MARK: - Preview
struct RootScreen_Previews: PreviewProvider {
    static var previews: some View {
        RootScreen()
    }
}

File: suc/kotlin/026.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "027*DocReader*en"
private const val UI_TYPE = "Productivity"
private const val STYLE_THEME = "Glassmorphic"
private const val LANG = "en"
private const val BASELINE_SIZE = "750x1334"

object AppTokens {
    object Colors {
        val primary = Color(0xFF38BDF8)
        val secondary = Color(0xFFA78BFA)
        val tertiary = Color(0xFF7DD3FC)
        val background = Color(0xFFF1F5F9)
        val surface = Color(0xFFFFFFFF)
        val surfaceVariant = Color(0xFFE2E8F0)
        val outline = Color(0xFFCBD5E1)
        val success = Color(0xFF22C55E)
        val warning = Color(0xFFF59E0B)
        val error = Color(0xFFEF4444)
        val onPrimary = Color(0xFFFFFFFF)
        val onSecondary = Color(0xFF1E1E1E)
        val onTertiary = Color(0xFF1E1E1E)
        val onBackground = Color(0xFF1E1E1E)
        val onSurface = Color(0xFF1E1E1E)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(12.dp)
        val large = RoundedCornerShape(20.dp)
    }
    object Spacing {
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 32.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.12f)
        val level2 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.18f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    tertiary = AppTokens.Colors.tertiary,
    onTertiary = AppTokens.Colors.onTertiary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RootScreen() {
    var scrollOffset by remember { mutableStateOf(Offset(0f, 0f)) }

    Scaffold(
        contentWindowInsets = WindowInsets(0),
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text("Glass Doc Reader", style = MaterialTheme.typography.displayLarge, color = AppTokens.Colors.onSurface) },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors(containerColor = AppTokens.Colors.background)
            )
        },
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(
                    Brush.verticalGradient(
                        listOf(
                            AppTokens.Colors.surfaceVariant.copy(alpha = 0.8f),
                            AppTokens.Colors.background,
                            AppTokens.Colors.surfaceVariant.copy(alpha = 0.8f)
                        )
                    )
                )
                .pointerInput(Unit) {
                    detectDragGestures { change, dragAmount ->
                        scrollOffset += dragAmount
                    }
                },
            contentAlignment = Alignment.Center
        ) {
            Surface(
                modifier = Modifier
                    .fillMaxWidth(0.9f)
                    .height(400.dp)
                    .offset(y = scrollOffset.y.dp / 5),
                color = AppTokens.Colors.surface.copy(alpha = 0.7f),
                shape = AppTokens.Shapes.large,
                shadowElevation = AppTokens.ElevationMapping.level2.elevation
            ) {
                Column(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(AppTokens.Spacing.lg),
                    verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
                ) {
                    Text("Document Reader", style = MaterialTheme.typography.titleMedium, color = AppTokens.Colors.primary)
                    repeat(8) {
                        Text(
                            text = "This is a sample line of text content number ${it + 1}.",
                            style = MaterialTheme.typography.bodyMedium,
                            color = AppTokens.Colors.onSurface
                        )
                    }
                }
            }
        }
    }

}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFF1F5F9)
@Composable
fun PreviewScreen() {
    AppTheme {
        RootScreen()
    }
}
